### ISSUE TITLE
invalid config.h if stdarg.h and varargs.h both missing

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12060]
Migrated from [rt.perl.org#112494](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=112494) (status was 'resolved')

Searchable as RT112494$

**Comment by p5pRT** [comment]
### From @nwc10

CompCert is a "verified compiler\, a high\-assurance compiler for almost all
of the ISO C90 / ANSI C language\, generating efficient code for the PowerPC\,
ARM and x86 processors\." \-\- http&#8203;://compcert\.inria\.fr/

So as it builds on Linux\, I decided to try it\. While building it I
discovered that "almost all" doesn't include variable argument lists\, which
is kind of a bugger here\, but I tried out running Configure with it to see
what happened\.

Well\, we get a config\.h\, but then we see this&#8203;:

In file included from perl\.h&#8203;:38&#8203;:0\,
&nbsp;                 from miniperlmain\.c&#8203;:40&#8203;:
config\.h&#8203;:2893&#8203;:3&#8203;: error&#8203;: invalid preprocessing directive \#I\_STDARG
config\.h&#8203;:2894&#8203;:3&#8203;: error&#8203;: invalid preprocessing directive \#I\_VARARGS
config\.h&#8203;:4400&#8203;:3&#8203;: error&#8203;: invalid preprocessing directive \#NEED\_VA\_COPY


Problem is that 3 variables in config\.sh are not set at all&#8203;:

$ egrep 'i\_stdarg|i\_varargs|va\_copy' config\.sh 
i\_stdarg=''
i\_varargs=''
need\_va\_copy=''

each should be 'define' or 'undef' \- eg&#8203;:

$ egrep 'i\_stdarg|i\_varargs|va\_copy' config\.sh  
i\_stdarg='define'
i\_varargs='undef'
need\_va\_copy='define'

The relevant output of Configure also looks screwball&#8203;:

\<stdarg\.h> found\.
 
\<varargs\.h> found\.
 
We'll include \<undef> to get va\_dcl definition\.
You don't have \<stdarg\.h> and \<stdlib\.h>\, so not checking for C99 variadic macro
s\.



The following patch fixes Configure to produce a valid config\.h
\(smoking as smoke\-me/no\-variable\-args \- don't merge this before 5\.16\.0\)

commit 0f0343205044d9c426ace82342a0831dbb80a74a
Author&#8203;: Nicholas Clark \<nick@&#8203;ccl4\.org>
Date&#8203;:   Tue Apr 17 12&#8203;:52&#8203;:44 2012 \+0200

&nbsp;    Generate a valid config\.h even if stdarg\.h and varargs\.h are both missing\.
&nbsp;    
&nbsp;    Whilst we can't build perl on such a crippled system\, there's no excuse for
&nbsp;    logical errors in config\.sh resulting in an invalid config\.h

<details><summary>Inline Patch</summary>

```diff
diff --git a/Configure b/Configure
index fdbbf20..fb0216b 100755
--- a/Configure
+++ b/Configure
@@ -10848,6 +10848,7 @@ chmod +x varargs
 : now check which varargs header should be included
 echo " "
 i_varhdr=''
+val=''
 case "$valstd" in
 "$define")
 	if `./varargs I_STDARG`; then
@@ -20248,6 +20249,7 @@ EOCP
 	$rm_try
 	;;
 *)	echo "You don't have <stdarg.h>, not checking for va_copy()." >&4
+	need_va_copy="$undef"
 	;;
 esac
 



```
</details>
With this, the build now fails cleanly, which is progress:

\`sh  cflags "optimize=''" perlmini\.o\`  \-DPERL\_IS\_MINIPERL \-DPERL\_EXTERNAL\_GLOB perlmini\.c
&nbsp;          CCCMD =  /usr/local/compcert\-1\.10/bin/ccomp \-DPERL\_CORE \-c \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64    
/usr/include/bits/sys\_errlist\.h&#8203;:28&#8203;: Warning&#8203;:
&nbsp;  redefinition of 'sys\_errlist' with incompatible type\.
perl\.c&#8203;:1628&#8203;: Warning&#8203;: argument \#1 of function call has type int \*
&nbsp;  instead of the expected type time\_t \*\.
perl\.c&#8203;:4019&#8203;: Error&#8203;: Unbound identifier '\_\_builtin\_va\_start'\.
Fatal error\.
1 error detected\.
make&#8203;: \*\*\* \[perlmini\.o\] Error 2


However\, as we can't build perl without variable argument lists\, I'm wondering
whether this bit of Configure&#8203;:

case "$val" in
''\)
echo "I could not find the definition for va\_dcl\.\.\. You have problems\.\.\." >&4
&nbsp;    val="$undef"; set i\_stdarg; eval $setvar
&nbsp;    val="$undef"; set i\_varargs; eval $setvar
&nbsp;    ;;

should either abort unconditionally\, or ask to abort\, default 'y'\, as the
build error is cryptic and unhelpful\, and doesn't really convey the true
cause of "your compiler isn't sufficiently a C compiler"\.

Nicholas Clark


Summary of my perl5 \(revision 5 version 15 subversion 8\) configuration&#8203;:   
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=3\.0\.0\-13\-generic\, archname=x86\_64\-linux
&nbsp;    uname='linux aranea 3\.0\.0\-13\-generic \#22\-ubuntu smp wed nov 2 13&#8203;:27&#8203;:26 utc 2011 x86\_64 x86\_64 x86\_64 gnulinux '
&nbsp;    config\_args='\-Dusedevel \-Dcc=/usr/local/compcert\-1\.10/bin/ccomp \-Dld=/usr/local/compcert\-1\.10/bin/ccomp \-Doptimize=  \-Dlibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc \-de'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=undef
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='/usr/local/compcert\-1\.10/bin/ccomp'\, ccflags ='\-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize=' '\,
&nbsp;    cppflags='\-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion=''\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=undef\, longlongsize=\, d\_longdbl=undef\, longdblsize=
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='ld'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib /usr/lib /lib64 /usr/lib64
&nbsp;    libs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.13'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_none\.xs\, dlext=none\, d\_dlsymun=undef\, ccdlflags=''
&nbsp;    cccdlflags=''\, lddlflags=''

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Nicholas Clark wrote&#8203;:
>However\, as we can't build perl without variable argument lists\, I'm wondering
\.\.\.
>should either abort unconditionally\, or ask to abort\,

Yeah\, you may as well bomb out early\.  "That's not a C compiler"\.
\<stdarg\.h> is one of the headers that in a formal sense is considered
to be part of the language\, rather than part of the library; supplying
it is the compiler's responsibility\.

\-zefram

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, Apr 17\, 2012 at 12&#8203;:24&#8203;:25PM \+0100\, Zefram wrote&#8203;:
> Nicholas Clark wrote&#8203;:
> >However\, as we can't build perl without variable argument lists\, I'm wondering
> \.\.\.
> >should either abort unconditionally\, or ask to abort\,
> 
> Yeah\, you may as well bomb out early\.  "That's not a C compiler"\.
> \<stdarg\.h> is one of the headers that in a formal sense is considered
> to be part of the language\, rather than part of the library; supplying
> it is the compiler's responsibility\.

Gosh\, Wikipedia fail\. http&#8203;://en\.wikipedia\.org/wiki/Stdarg\.h
a\) doesn't say that it's mandatory
b\) doesn't list which headers are

\(I thought that there are 4 for C89\, but I can't remember which\.
And I can't remember the terms for the language vs language\+library\.
"hosted" was the latter?\)

Nicholas Clark

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Nicholas Clark wrote&#8203;:
>\(I thought that there are 4 for C89\, but I can't remember which\.

Yes\.  \<float\.h>\, \<limits\.h>\, \<stdarg\.h>\, \<stddef\.h>\.  \<iso646\.h> was
added in an amendment\, and C90 added a few more such as \<stdint\.h>\.

>And I can't remember the terms for the language vs language\+library\.
>"hosted" was the latter?\)

"Freestanding implementation" and "hosted implementation"\.

\-zefram

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Apr 17 04&#8203;:36&#8203;:42 2012\, zefram@&#8203;fysh\.org wrote&#8203;:
> Nicholas Clark wrote&#8203;:
> >\(I thought that there are 4 for C89\, but I can't remember which\.
> 
> Yes\.  \<float\.h>\, \<limits\.h>\, \<stdarg\.h>\, \<stddef\.h>\.  \<iso646\.h> was
> added in an amendment\, and C90 added a few more such as \<stdint\.h>\.
> 
> >And I can't remember the terms for the language vs language\+library\.
> >"hosted" was the latter?\)
> 
> "Freestanding implementation" and "hosted implementation"\.
> 
> \-zefram
> 

Discussion in this RT petered out last April\.

Is there a bug in Perl here?  Or a bug in the CompCert compiler?  \(Or both?\)

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @arc

James E Keenan via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> Discussion in this RT petered out last April\.
>
> Is there a bug in Perl here?  Or a bug in the CompCert compiler?  \(Or both?\)

To summarise the earlier discussion&#8203;:

1\. The compiler in question compiles something sufficiently unlike C
\(of any vintage\) that Perl can't be built with it

2\. One manifestation of that is that neither the standard \<stdarg\.h>
nor pre\-standard \<varargs\.h> headers are available

3\. An infelicity of Perl's build system means that\, when \(2\) is true\,
the config\.h we generate is syntactically invalid

4\. Nick has a patch to ensure that config\.h is syntactically correct regardless

5\. There seems to be consensus that it would also be useful for
Configure to bail out when encountering such a C compiler

Nick's original patch looks sane to me as a way of ensuring that
config\.h is syntactically valid in such situations\, and is still
available as the tip of smoke\-me/no\-variable\-args\, for which the
results are here&#8203;:

http&#8203;://perl\.develop\-help\.com/?b=smoke\-me/no\-variable\-args

There's quite a lot of red on that page\, but I can't see anything
there that looks like it might be attributable to the change\.  I don't
know what our procedure is for dealing with smoke\-me failures like
that\.  If we should go ahead with it anyway\, I'm happy to \(rebase and\)
merge Nick's patch to blead; Nick\, want me to do so?

As for having Configure bail out when the compiler doesn't support
varargs\, I've just pushed a patch along those lines to
smoke\-me/no\-varargs\.

\-\- 
Aaron Crane \*\* http&#8203;://aaroncrane\.co\.uk/

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon Feb 04 09&#8203;:30&#8203;:21 2013\, arc wrote&#8203;:
> James E Keenan via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > Discussion in this RT petered out last April\.
> >
> > Is there a bug in Perl here?  Or a bug in the CompCert compiler?  \(Or
> > both?\)
> 
> To summarise the earlier discussion&#8203;:
> 
> 1\. The compiler in question compiles something sufficiently unlike C
> \(of any vintage\) that Perl can't be built with it
> 
> 2\. One manifestation of that is that neither the standard \<stdarg\.h>
> nor pre\-standard \<varargs\.h> headers are available
> 
> 3\. An infelicity of Perl's build system means that\, when \(2\) is true\,
> the config\.h we generate is syntactically invalid
> 
> 4\. Nick has a patch to ensure that config\.h is syntactically correct
> regardless
> 
> 5\. There seems to be consensus that it would also be useful for
> Configure to bail out when encountering such a C compiler
> 
> Nick's original patch looks sane to me as a way of ensuring that
> config\.h is syntactically valid in such situations\, and is still
> available as the tip of smoke\-me/no\-variable\-args\, for which the
> results are here&#8203;:
> 
> http&#8203;://perl\.develop\-help\.com/?b=smoke\-me/no\-variable\-args
> 
> There's quite a lot of red on that page\, but I can't see anything
> there that looks like it might be attributable to the change\.  I don't
> know what our procedure is for dealing with smoke\-me failures like
> that\.  If we should go ahead with it anyway\, I'm happy to \(rebase and\)
> merge Nick's patch to blead; Nick\, want me to do so?
> 
> As for having Configure bail out when the compiler doesn't support
> varargs\, I've just pushed a patch along those lines to
> smoke\-me/no\-varargs\.

Aaron\, Nicholas&#8203;:

There has been no movement on this ticket in more than a year\-and\-a\-half\.  Is there some patch you would like us to consider?  \(Otherwise\, is the ticket closable?\)

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @arc

The "this isn't a C compiler" patch to Configure was merged in February 2013\, as b95f969b03c79d1a9cda098c611d894dd874b35a\. I have come to the conclusion that it makes Nick's original patch unnecessary \(since keeping config\.h well\-formed even when the compiler is inadequate is of essentially no value when we already refuse to use such compilers\)\. I'm therefore resolving this ticket\.

James\, thanks for chasing this up\.

\-\- 
Aaron Crane \*\* http&#8203;://aaroncrane\.co\.uk/

**Comment by p5pRT** [comment]
@arc - Status changed from 'open' to 'resolved'
